frontend/src/hooks/useValidation.js

```diff
@@ -0,0 +1,298 @@
+import { useState } from "react";
+import { omit, mapValues, minBy, maxBy } from "lodash";
+import { differenceInCalendarDays } from "date-fns";
+
+// Validation functions for the SubmitForm component
+export default function useValidation(dates, notes, season) {
+  // Validation errors
+  const [errors, setErrors] = useState({});
+  // @TODO: Track if the form has been submitted, and validate on change after that
+  const [formSubmitted, setFormSubmitted] = useState(false);
+
+  // Add field key and error message to `errors`
+  function addError(fieldId, message) {
+    setErrors((prevErrors) => ({
+      ...prevErrors,
+      [fieldId]: message,
+    }));
+    return false;
+  }
+
+  // Remove field key from `errors`
+  function clearError(fieldId) {
+    setErrors((prevErrors) => omit(prevErrors, fieldId));
+  }
+
+  // Returns true if innerStart and innerEnd are within outerStart and outerEnd
+  function checkWithinRange(outerStart, outerEnd, innerStart, innerEnd) {
+    return outerStart <= innerStart && outerEnd >= innerEnd;
+  }
+
+  // Calculates the extents of each date type for each campsite grouping
+  function getDateExtents(datesObj = dates) {
+    const mapped = mapValues(datesObj, (campsiteDates) =>
+      mapValues(campsiteDates, (dateTypeDates) => {
+        // Get the dateRange with the earliest start date for this campground & date type
+        const minDateRange = minBy(
+          dateTypeDates,
+          (dateRange) => new Date(dateRange.startDate),
+        );
+        const minDate = minDateRange?.startDate
+          ? new Date(minDateRange?.startDate)
+          : null;
+
+        // Get the dateRange with the latest end date for this campground & date type
+        const maxDateRange = maxBy(
+          dateTypeDates,
+          (dateRange) => new Date(dateRange.endDate),
+        );
+        const maxDate = maxDateRange?.endDate
+          ? new Date(maxDateRange.endDate)
+          : null;
+
+        return { minDate, maxDate };
+      }),
+    );
+
+    return mapped;
+  }
```

Comment:
I know you mentioned this is on TODO to account for gaps for non-continuous date ranges, but I think we should include it on this PR. Probably just need to check for each reservation date range one at a time and make sure that it's covered.

Reply:
I just pushed an update that does a more robust check on the date ranges: 
1. sort the dates chronologically
2. combine any overlaps
3. _then_ make sure all reservation ranges are within an operating date range